include/wil/registry.h

```diff
@@ -17,16 +17,3139 @@
 
 #include <winreg.h>
 #include <new.h> // new(std::nothrow)
-#include "resource.h" // unique_hkey
+#include "registry_helpers.h"
+#include "resource.h"
+
+#ifndef __WIL_WINREG_
+#error This is required for ::wil::unique_hkey support
+#endif
+
+// wil registry does not require the use of the STL or C++ exceptions
+// wil registry natively supports std::vector and std::wstring when preferring those types
+// wil registry uses the __WIL_WINREG_STL define to enable support for wil::shared_* types (defined in resource.h)
+// include sddl.h if wanting to pass security descriptor strings when creating keys
 
 namespace wil
 {
-    //! The key name includes the absolute path of the key in the registry, always starting at a
-    //! base key, for example, HKEY_LOCAL_MACHINE.
-    size_t const max_registry_key_name_length = 255;
+    namespace reg
+    {
+        /**
+         * \brief Helper function to translate registry return values if the value was not found
+         * \param hr HRESULT to test from registry APIs
+         * \return boolean if the HRESULT indicates the registry value was not found
+         */
+        constexpr bool is_hresult_not_found(HRESULT hr) WI_NOEXCEPT
+        {
+            return hr == HRESULT_FROM_WIN32(ERROR_FILE_NOT_FOUND);
+        }
+
+        /**
+         * \brief Helper function to translate registry return values if the buffer was too small
+         * \param hr HRESULT to test from registry APIs
+         * \return boolean if the HRESULT indicates the buffer was too small for the value being read
+         */
+        constexpr bool is_hresult_buffer_too_small(HRESULT hr) WI_NOEXCEPT
+        {
+            return hr == HRESULT_FROM_WIN32(ERROR_MORE_DATA);
+        }
+
+#if defined(WIL_ENABLE_EXCEPTIONS)
+        /**
+         * \brief Opens a new HKEY to the specified path
+         *        see RegOpenKeyExW
+         * \param key An opened registry key, or fixed registry key as the base key, from which to append the path
+         * \param path A string to append to the HKEY to attempt to open
+         *        can be nullptr if not needed
+         * \param access The requested access desired for the opened key
+         * \return A unique_hkey containing the resulting opened HKEY
+         * \exception std::exception will be thrown on failure (including wil::ResultException)

```

Comment:
I'd remove any occurrence of a line like this---exceptions are implied for functions that are not noexcept.

Reply:
my response is that it is definitely not implicit to all C / C++ developers.
I would prefer to have an extra one line of verbosity so that the exception behavior is explicit.

Reply:
I understand that reasoning, but I don't think that's the correct approach here for a few reasons:

1. We don't do it elsewhere, so such a dev would now be confused by other parts of our code where we don't have these lines.
2. It _really_ is redundant documentation. Every function in C++, including in the STL, is assumed to throw unless explicitly marked with `noexcept`. Adding documentation here is a red flag to devs who know that---why is it documented here specifically?
3. It doesn't show up in IDEs anyway. Visual Studio does not show this line in autocomplete, so it doesn't really help folks.

Documenting things redundantly adds maintenance cost to this code, and I don't think it offers any concrete benefit. We're not trying to document WIL for folks who don't know the STL exception policy, IMO.

Reply:
I still don't think we should explicitly document exceptions for my reasons above, but I do not think it's a blocker for this PR.